common/algorithms/murmur-hash.cpp

```diff
@@ -0,0 +1,83 @@
+//  Modified by LLC «V Kontakte», 2025 February 25
+//
+//  This file is part of the GNU C Library.
+//  Copyright (C) 2002-2024 Free Software Foundation, Inc.
+//  This program is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+//  GNU General Public License for more details.
+//
+//  You should have received a copy of the GNU General Public License
+//  along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+#include "common/algorithms/hashes.h"
+
+#include <cstddef>
+#include <cstdint>
+
+namespace {
+
+size_t unaligned_load(const char *p) noexcept {
+  size_t result = 0;
+  __builtin_memcpy(&result, p, sizeof(result));
+  return result;
+}
+
+// Loads n bytes, where 1 <= n < 8.
+size_t load_bytes(const char *p, int n) noexcept {
+  size_t result = 0;
+  --n;
+  do {
+    result = (result << 8) + static_cast<unsigned char>(p[n]);
+  } while (--n >= 0);
+  return result;
+}
+
+size_t shift_mix(size_t v) noexcept {
+  return v ^ (v >> 47);
+}
+
+} // namespace
+
+namespace vk {
+
+// MurMur hash function was taken from libstdc++
+template<>
+uint64_t murmur_hash<uint64_t>(const void *ptr, size_t len, size_t seed) noexcept {
```

Comment:
Can't it be defined as `constexpr`?

Reply:
It'd force us to define it in header. Anyway, there is not much point in doing this as most of `get_hash` methods are `virtual` which means we can't make them `constexpr`